libp2p/network/swarm.py

```diff
@@ -288,19 +271,46 @@ def add_router(self, router: IPeerRouting) -> None:
         # Store muxed_conn with peer id
         self.connections[muxed_conn.peer_id] = swarm_conn
         # Call notifiers since event occurred
-        for notifee in self.notifees:
-            # TODO: Call with other type of conn?
-            await notifee.connected(self, muxed_conn)
+        self.notify_connected(swarm_conn)
         await swarm_conn.start()
         return swarm_conn
 
     def remove_conn(self, swarm_conn: SwarmConn) -> None:
         """
         Simply remove the connection from Swarm's records, without closing the connection.
         """
-        peer_id = swarm_conn.conn.peer_id
+        peer_id = swarm_conn.muxed_conn.peer_id
         if peer_id not in self.connections:
             return
         # TODO: Should be changed to remove the exact connection,
         #   if we have several connections per peer in the future.
         del self.connections[peer_id]
+
+    # Notifee
+
+    # TODO: Remeber the spawn notifying tasks and clean them up when closing.
+
+    def register_notifee(self, notifee: INotifee) -> None:
+        """
+        :param notifee: object implementing Notifee interface
+        :return: true if notifee registered successfully, false otherwise
+        """
+        self.notifees.append(notifee)
+
+    def notify_opened_stream(self, stream: INetStream) -> None:
+        asyncio.gather(
```

Comment:
something to think about here is what happens to these tasks after they are started by `asyncio.gather`.

afaiui, the tasks will start and then run concurrently w/ other tasks and hopefully these tasks complete. if we have to cancel the notification, we won't have a way to do it

not sure this change blocks merging in this PR as to my knowledge we aren't using the notifications system very much right now... shall we move this to an issue?

Reply:
Yes, you're right. I opened https://github.com/libp2p/py-libp2p/issues/309 to track this one.